### ISSUE TITLE
Add more unit tests for WebSockets

### ISSUE THREAD

**Issue by gavv** [gavv/httpexpect#70]
WebSocket support is covered with a set of [integration tests](https://github.com/gavv/httpexpect/blob/master/e2e_websocket_test.go), but only a few unit tests: [1](https://github.com/gavv/httpexpect/blob/master/websocket_test.go), [2](https://github.com/gavv/httpexpect/blob/master/websocket_message_test.go).

The test coverage is incomplete, see `websocket*` files here: https://coveralls.io/github/gavv/httpexpect?branch=master

It would be nice to add more unit tests and increase test coverage. In particular, most assertion *failures* and some message types are not covered currently.

**Comment by jaibhavaya** [comment]
Is this available?  If so I'd love to take a crack at it.

**Comment by gavv** [comment]
Websocket methods still not covered with unit tests:

* WriteBytesBinary
* WriteBytesText
* WriteText

**Comment by djanda97** [comment]
@gavv I don't have much experience with Go or WebSockets but I will take a shot at implementing unit tests for the remaining methods.

**Comment by djanda97** [comment]
@gavv To clarify my understanding, should the unit tests for the remaining WebSocket methods be similar to the method calls of WriteBytesBinary, WriteBytesText, and WriteText in [e2e_websocket_test.go](https://github.com/gavv/httpexpect/blob/master/e2e_websocket_test.go) but with the addition of test cases where an error would be produced?

**Comment by gavv** [comment]
@djanda97 Hi, I think we should wait until #75 is merged to avoid merge conflicts (@nio6412 is working on it).

> should the unit tests for the remaining WebSocket methods be similar to the method calls of WriteBytesBinary, WriteBytesText, and WriteText in e2e_websocket_test.go but with the addition of test cases where an error would be produced?

Not exactly, e2e_websocket_test.go implements integration (e2e stands for "end-to-end") tests with a real websocket connection. And here we need unit tests, which will require mocking websocket conn, which is not possible right now but likely will be implemented in #75 as suggested here: https://github.com/gavv/httpexpect/pull/75#discussion_r437904235

**Comment by djanda97** [comment]
@gavv Thank you for clarifying the distinction between e2e vs unit tests. I'll check back in once #75 is merged!

**Comment by unbeatable01** [comment]
I'd like to work on this, could you assign it to me?

**Comment by jan-carreras** [comment]
Just to clarify. 

1. Are you only interested in increasing the _missing_ coverage for both Unit and E2E tests? 

2. The unit tests on WebSocket are lacking on assertions and the code could be changed to[1]. Would you be interested in that or do you think the current unit tests are good enough?


[1] From this code that _only_ adds coverage but does not _assert_ anything:

```
ws.WithReadTimeout(0)
ws.WithoutReadTimeout()
ws.WithWriteTimeout(0)
ws.WithoutWriteTimeout()
```

To this code:

```	
timeout := time.Millisecond
ws.WithReadTimeout(timeout)
assert.Equal(t, timeout, ws.readTimeout)
ws.WithoutReadTimeout()
assert.Equal(t, noDuration, ws.readTimeout)
ws.WithWriteTimeout(timeout)
assert.Equal(t, timeout, ws.writeTimeout)
ws.WithoutWriteTimeout()
assert.Equal(t, noDuration, ws.writeTimeout)`
```

**Comment by gavv** [comment]
Thanks everyone, I've finally merged #75 by @nio6412 and the coverage is pretty good now. 

There are still a few uncovered lines in websocket_dialer and websocket_message, but that's not critical. I'm going to close this issue, but feel free to let me know if someone would like to add a few more tests.

**Comment by gavv** [comment]
@jan-carreras 

> Are you only interested in increasing the missing coverage for both Unit and E2E tests?

Ideally unit tests should cover everything even without e2e tests, but that's not a strict requirement.

> The unit tests on WebSocket are lacking on assertions and the code could be changed to[1]. Would you be interested in that or do you think the current unit tests are good enough?

After merging the PR, we have most of the needed assertions now. Though probably something can be improved.

>  From this code that only adds coverage but does not assert anything:

That specific test just checks that the methods wont panic when called on an invalid/failed object. More specific checks are now present in other tests.
